libarchive/archive_read_disk_posix.c

```diff
@@ -668,14 +668,12 @@ setup_suitable_read_buffer(struct archive_read_disk *a)
 		 */
 		s = (uintptr_t)cf->allocation_ptr;
 		s %= xfer_align;
-		if (s > 0)
-			s = xfer_align - s;
 
 		/*
 		 * Set a read buffer pointer in the proper alignment of
 		 * the current filesystem.
 		 */
-		cf->buff = cf->allocation_ptr + s;
+		cf->buff = cf->allocation_ptr + xfer_align - s;
```

Comment:
It's possible to avoid the branch above if you assume `xfer_align` is a power of two.  However, the change you've made here actually wastes `xfer_align` bytes in the case where the original allocation is already aligned (which should be the common case).

Reply:
Oh.  Yes, actually you can do this correctly, but you need an additional modulo operation:
```
cf->buff = cf->allocation_ptr + ((xfer_align - s) % xfer_align)
```